feat: Create rfc documents from rfc index updates

jennifer-richards opened the pull request.

jennifer-richards committed:
style: Run Black on update_docs_from_rfc_index()

jennifer-richards committed:
chore: Add type annotation to complex return value

jennifer-richards committed:
feat: Use rfc doc type when updating from rfc index

jennifer-richards committed:
style: Run Black and use f-strings / str.format()

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -367,80 +370,140 @@ def update_docs_from_rfc_index(index_data, errata_data, skip_older_than_date=Non
+            try:
+                draft = Document.objects.get(name=draft_name)
+            except Document.DoesNotExist:
+                pass
```
We should probably log this - the RPC thinks the doc came from a draft, but we can't find it. That's probably a real error.

jennifer-richards replied:
will do

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -367,80 +370,140 @@ def update_docs_from_rfc_index(index_data, errata_data, skip_older_than_date=Non
+                    draft_changes.append(change)
+                    rfc_changes.append(change)
+
+                # Ensure draft is in the correct iesg and stream states
+                for t in (
+                    "draft-iesg",
+                    "draft-stream-iab",
+                    "draft-stream-irtf",
+                    "draft-stream-ise",
+                ):
+                    prev_state = draft.get_state(t)
+                    if prev_state is not None:
+                        if prev_state.slug not in ("pub", "idexists"):
+                            new_state = State.objects.select_related("type").get(
+                                used=True, type=t, slug="pub"
+                            )
+                            draft.set_state(new_state)
+                            draft_changes.append(
+                                f"changed {new_state.type.label} to {new_state}"
+                            )
+                            e = update_action_holders(draft, prev_state, new_state)
+                            if e:
+                                draft_events.append(e)
+                    elif t == "draft-iesg":
+                        draft.set_state(
+                            State.objects.get(type_id="draft-iesg", slug="idexists")
+                        )
```
This doesn't make sense to me.
Instead, verify that the draft stream matches the rfc stream, and then verify the states only for that stream.

rjsparks replied:
I see below that this is what the old code did, but I think that was wrong.

jennifer-richards committed:
refactor: Consolidate draft handling

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -367,80 +370,140 @@ def update_docs_from_rfc_index(index_data, errata_data, skip_older_than_date=Non
         if title != doc.title:
             doc.title = title
-            changes.append("changed title to '%s'" % doc.title)
+            rfc_changes.append(f"changed title to '{doc.title}'")
```
i'm a little itchy at "changed" here. I wonder if "set" would be better? Since doc is the new RFC document, you're not checking to see if the RFC had a different title than the draft did. Perhaps the gearing should change to make and report on _that_ instead? Or am I misreading?

jennifer-richards replied:
You've drawn my attention to a bigger bug. These fields should be populated from the draft to start with. With that done, I *think* the "change" language will continue to make sense.

jennifer-richards replied:
(though I can make it choose "set" vs "change" based on the value of `created_rfc` if you prefer that)

jennifer-richards replied:
(continuing to talk to myself here) Never mind - we aren't guaranteed to have a draft to copy the fields from. I'll have it report "set" vs "changed" as appropriate.

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -467,35 +530,24 @@ def update_docs_from_rfc_index(index_data, errata_data, skip_older_than_date=Non
                 #   l = DocAlias.objects.filter(name__startswith="rfc", docs__docalias__name=x.lower())
-                #else:
+                # else:
                 l = Document.objects.filter(name=x.lower())
```
I wonder if we should be filtering this to doc type rfc?

jennifer-richards replied:
I'd say so

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -504,56 +556,96 @@ def parse_relation_list(l):
 
-        doc_errata = errata.get('RFC%04d'%rfc_number, [])
-        all_rejected = doc_errata and all( er['errata_status_code']=='Rejected' for er in doc_errata )
+        doc_errata = errata.get("RFC%04d" % rfc_number, [])  # rfc10k problem here
```
Lets get rid of the %04 here now?

rjsparks replied:
do errata searches actually take leading zeros?

jennifer-richards replied:
I can make this handle leading zeroes or not - that'll avoid having to closely coordinate the change.

rjsparks commented on ietf/sync/rfceditor.py:
```diff
@@ -504,56 +556,96 @@ def parse_relation_list(l):
+                )
+            )
+            doc.save_with_history(rfc_events)
+            yield rfc_changes, doc, rfc_published  # yield changes to the RFC
```
will this yielding do the right thing if there are both draft_changes and rfc_changes?

jennifer-richards replied:
Yes:

jennifer-richards replied:
(this is also confirmed *in situ* by `test_rfc_index()`)

jennifer-richards replied:
(rather, will be when I commit that piece)

jennifer-richards committed:
fix: Only match draft docs to rfc index draft name

jennifer-richards committed:
chore: Eliminate python lint

jennifer-richards committed:
test: Update test_rfc_index()

jennifer-richards committed:
test: Confirm changes are returned for draft and rfc

jennifer-richards committed:
fix: Only consider rfcs for updates/obsoletes targets

jennifer-richards committed:
feat: Log warning if rfc index refers to unknown draft

jennifer-richards committed:
feat: Set more fields on rfc doc

jennifer-richards committed:
feat: Set published state on rfc doc

jennifer-richards committed:
feat: describe change as "set" or "changed" as appropriate

jennifer-richards committed:
fix: Check state correctly

jennifer-richards committed:
test: Test more fields on the rfc doc

jennifer-richards committed:
test: Remove python lint

jennifer-richards committed:
fix: Drop leading 0s from RFC doc-id

jennifer-richards committed:
test: Test behavior of verified-errata tag

Also add comments flagging the near-miss between doc-id
and RFC number in the test_rfc_index() test data. It's
unclear whether these were intended to match, but the
test was self-consistent so I am not changing it.

jennifer-richards committed:
refactor: Refactor handling of draft stream states

jennifer-richards committed:
fix: Set draft-iesg state to "pub" when a draft is published

jennifer-richards committed:
fix: Use correct stream slugs

jennifer-richards committed:
test: Don't mix stream states in test_rfc_index()

Make the test a better simulation of actual practice.

jennifer-richards committed:
test: Check that no logs are emitted in test_rfc_index

jennifer-richards commented:
I believe I addressed all your comments, plus some other changes. I suggest looking closely at 7535df3eeb67dddcb9756f17761828e7d98c27bf which brings more fields over from the draft to the new rfc document (these are also the fields set on the rfc document in the `...create_rfc_documents.py` migration). The refactored stream state handling is also worth checking in detail.

jennifer-richards committed:
fix: Restore old handling of draft-iesg state

Missing state -> idexists, idexists -> idexists, anything
else becomes pub.
